File: src/II040.py
from typing import List


class Solution:
    def maximalRectangle(self, matrix: List[str]) -> int:
        m = len(matrix)
        if m == 0:
            return 0
        n = len(matrix[0])
        if n == 0:
            return 0
        ret = 0
        # cnt = [[0]*n]*m # 注意这样构建出来的二维数组，各行是引用关系，都一样
        cnt = [[0]*n for _ in range(m)]  # 需要这样构建二维数组！
        for i in range(m):
            for j in range(n):
                if matrix[i][j] == '1':
                    cnt[i][j] = 1 if j == 0 else cnt[i][j-1]+1
        for j in range(n):
            st = []
            left = [0]*m
            for i in range(m):
                while st and cnt[st[len(st)-1]][j] >= cnt[i][j]:
                    st.pop()
                left[i] = -1 if not st else st[len(st)-1]
                st.append(i)
            st = []
            for i in reversed(range(m)):
                while st and cnt[st[len(st)-1]][j] >= cnt[i][j]:
                    st.pop()
                right = m if not st else st[len(st)-1]
                st.append(i)
                ret = max(ret, (right-left[i]-1)*cnt[i][j])
        return ret


matrix = ["10100", "10111", "11111", "10010"]
print(Solution().maximalRectangle(matrix))

File: src/II040.cpp
#include <iostream>
#include <stack>
#include <string>
#include <utility>
#include <vector>

using namespace std;

class Solution {
  vector<int> cnt;
  int getLineRes() {
    stack<int> st;
    st.push(-1);
    cnt.push_back(0);
    int n = cnt.size(), res = 0;
    for (int i = 0; i < n; ++i) {
      while (st.size() > 1 && cnt[st.top()] > cnt[i]) {
        int height = cnt[st.top()];
        st.pop();
        int width = i - st.top() - 1;
        res = max(res, height * width);
      }
      st.push(i);
    }

    cnt.pop_back();
    return res;
  }

public:
  int maximalRectangle(vector<string> &matrix) {
    if (matrix.empty())
      return 0;
    int n = matrix[0].size(), res = 0;
    cnt = vector<int>(n);
    for (auto line : matrix) {
      for (int i = 0; i < n; ++i) {
        cnt[i] = line[i] == '0' ? 0 : cnt[i] + 1;
      }
      res = max(res, getLineRes());
    }
    return res;
  }
};

int main() {
  Solution solu;
  vector<string> test{"10100", "10111", "11111", "10010"};
  cout << solu.maximalRectangle(test) << endl;

  return 0;
}
